src/python/pants/core/util_rules/partitions.py

```diff
@@ -204,3 +228,78 @@ async def partitioner(
         )
 
     return collect_rules(locals())
+
+
+@memoized
+def _partition_per_input_field_sets_partitioner_rules(cls) -> Iterable:
+    """Returns a rule that implements a "partitioner" for `PartitionFieldSetsRequest`, which returns
+    a single-element partition per input."""
+
+    @rule(
+        _param_type_overrides={
+            "request": cls.PartitionRequest,
+            "subsystem": cls.tool_subsystem,
+        }
+    )
+    async def partitioner(
+        request: _PartitionFieldSetsRequestBase, subsystem: SkippableSubsystem
+    ) -> Partitions[FieldSet, Any]:
+        return (
+            Partitions()
+            if subsystem.skip
+            else Partitions(Partition((field_set,), None) for field_set in request.field_sets)
+        )
+
+    return collect_rules(locals())
+
+
+@memoized
+def _partition_per_input_field_sets_by_file_partitioner_rules(cls) -> Iterable:
+    """Returns a rule that implements a "partitioner" for `PartitionFieldSetsRequest`, which returns
+    a single-element partition per input."""
+
+    # NB: This only works if the FieldSet has a single `SourcesField` field. We check here for
+    # a better user experience.
+    sources_field_name = None
+    for fieldname, fieldtype in _get_field_set_fields(cls.field_set_type).items():
+        if issubclass(fieldtype, SourcesField):
+            if sources_field_name is None:
+                sources_field_name = fieldname
+                break
+            raise TypeError(
+                f"Type {cls.field_set_type} has multiple `SourcesField` fields."
+                + " Pants can't provide a default partitioner."
+            )
+    else:
+        raise TypeError(
+            f"Type {cls.field_set_type} has does not have a `SourcesField` field."
+            + " Pants can't provide a default partitioner."
+        )
```

Comment:
naughty naughty. Split this into a helper, we shouldn't be duplicating with `_single_partition_field_sets_by_file_partitioner_rules` :wink:

Reply:
Fair enough 😄 done!